#169817550 Manager can get notification on new trip

victor-abz opened the pull request:
Manager can opt-in to certain notification mode(email / in-app)
Manager can opt-out to certain notification mode

[Starts #169817550]
#### What does this PR do?
User can get notification
#### Description of Task to be completed?
- Manager can opt-in to certain notification mode(email / in-app)
- Manager can opt-out to certain notification mode
- Manager can get notification by preferred mode when a new trip created
#### How should this be manually tested?
go to terminal and do `git clone https://github.com/andela/the_spinners-backend`
switch to a branch by `git checkout ft-notification-on-new-travel-169817550`
run `npm run dev`
##### visit the following API in postman
###### To set preferred mode
PUT`/api/user/notification/set/email-notification` to set receive notification by email
PUT `/api/user/notification/set/in-app-notification` to receive notification in-app
###### To remove preferred mode
PUT `/api/users/notification/unset/email-notification` to unsubscribe from receiving notification by email
PUT `/api/user/notification/unset/in-app-notification` to unsubscribe from notification in-app
###### To receive sample notification
create a new trip by visiting POST `/api/one-way-trips`, you should get a notification in email.
to get all notification for a logged in user visit `/api/user/notification`

more on testing can be seen in our documentation [https://snippers-backend-stage.herokuapp.com/docs](https://snippers-backend-stage.herokuapp.com/docs)
#### Any background context you want to provide?
#### What are the relevant pivotal tracker stories?
#169817550 
#### Screenshots (if appropriate)
<img width="893" alt="Screen Shot 2020-01-01 at 23 45 30" src="https://user-images.githubusercontent.com/52409462/71646418-0ae8f080-2cf1-11ea-85b1-34f9abe5ac9e.png">

#### Questions:

higustave-ops commented on src/helpers/notifications/trip-notification.js:
```diff
@@ -0,0 +1,72 @@
+        type: 'new_request',
+        userId: id,
+      });
+      delete inAppNotification.dataValues.isRead;
```
Can you use `rest param destructuring` instead of `delete`

higustave-ops commented on src/services/send-email.service.js:
```diff
@@ -7,12 +7,11 @@ import JwtService from './jwt.service';
    * @param {body} emailBody description of the email
    * @returns {sendMessage} this function send an email
    */
-  static sendGridEmail(email, token, emailSubject, emailBody) {
```
Is our features going to work as they used to, because this method was for all of us and now, you removed the token `arg`. Need explanation. Thanks

victor-abz replied:
the token is not used anywhere in the body. s

higustave-ops replied:
Did you check that other features that were depending on this method are working. For example mine of user account email verification? The email sent should contain a token> Try with your email and see if it works then will be ok. Thanks

higustave-ops reviewed (approved):
LGTM

erickyvand commented on src/helpers/validateErrorHandler.js:
```diff
@@ -0,0 +1,18 @@
+import ResponseService from '../services/response.service';
+
+const validateErrorHandler = (schema, body, res, next) => {
```
Why did you put validation in the helpers folder while we have a validations folder?

victor-abz replied:
This is a helper function which brings all errors from joi at once. I saw we all had to copy it and that would cause duplicates as code grows. I call it to generate the error without rewrite it.

erickyvand commented:
LGTM

adesege commented on src/app.js:
```diff
@@ -3,14 +3,19 @@ import express from 'express';
 
 const app = express();
 app.use(morgan('dev'));
+app.use(express.static(`${__dirname}/public/`));
```
Is there a reason for adding this line?

adesege commented on src/controllers/notification.controller.js:
```diff
@@ -0,0 +1,40 @@
+    // If request is emailNotification
+    await PreferenceService
+      .updatePreference({ userId: req.userData.id }, { emailNotification: req.value });
+    ResponseService.setSuccess(200, `Successfully updated your ${req.option} preferences.`);
```
`req.option` does not have a user-friendly name. Use a user-friendly name instead

adesege commented on src/helpers/notifications/trip-notification.js:
```diff
@@ -0,0 +1,75 @@
+import PreferenceService from '../../services/preference.service';
+
+/** Class representing a Notifications . */
+class TripNotification {
```
This should move to the services directory instead.

adesege commented on src/helpers/validateErrorHandler.js:
```diff
@@ -0,0 +1,18 @@
+import ResponseService from '../services/response.service';
+
+const validateErrorHandler = (schema, body, res, next) => {
```
`transformErrorHandler` is more succinct than `validateErrorHandler`.

adesege commented on src/helpers/validateErrorHandler.js:
```diff
@@ -0,0 +1,18 @@
+  if (error) {
+    const errors = [];
+    const { details } = error;
+    details.forEach(({ message }) => {
```
Use .map instead.

adesege commented on src/middlewares/auth.middleware.js:
```diff
@@ -11,6 +11,14 @@ const authMiddleware = {
     }
     next();
   },
+  doesHaveManager: async (req, res, next) => {
```
`doesHaveManager` is not a good function name. Ensure your function name has a verb and clearly describes what the function does.

adesege commented on src/middlewares/notification.middleware.js:
```diff
@@ -0,0 +1,19 @@
+import PreferenceService from '../services/preference.service';
+import ResponseService from '../services/response.service';
+
+const doesOptionExist = async (req, res, next) => {
```
`doesOptionExist` is not a good function name. Ensure your function name has a verb and clearly describes what the function does.

adesege commented on src/middlewares/notification.middleware.js:
```diff
@@ -0,0 +1,19 @@
+  const prefs = await PreferenceService.findPreferences({ userId: req.userData.id });
+  const { dataValues } = prefs;
+  Object.keys(dataValues)
+    .forEach((index) => {
```
You can use a combination of filter and map functions here.

adesege commented on src/migrations/20200103080032-create-preferences.js:
```diff
@@ -0,0 +1,33 @@
+    userId: {
+      type: Sequelize.INTEGER
+    },
+    emailNotification: {
```
Variables that contain a boolean as a value should be prefixed with either `is` or `has`

adesege commented on src/routes/index.js:
```diff
@@ -2,11 +2,13 @@ import express from 'express';
 app.use('/api/auth', authRoute);
 app.use('/api/', tripRoute);
 app.use('/api/', profileSettingsRoute);
+app.use('/api/', notificationRoute);
```
The context can be scoped to `/api/notifications`

victor-abz replied:
Yeah, it is /api/notifications except that the last part `/notification` is in the notification route file.

adesege commented on src/helpers/notifications/trip-notification.js:
```diff
@@ -0,0 +1,75 @@
+      const { isRead, updateAt, createdAt, ...notificationData } = data.dataValues;
+
+      // email nofitication
+      if (emailNotification === true) {
```
Add a test case for this line and line 67

adesege commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "liveServer.settings.port": 5501
```
You don't need to push this file.

adesege commented on index.html:
```diff
@@ -0,0 +1,62 @@
+<!DOCTYPE html>
```
What does this HTML file do?

adesege commented on src/controllers/notification.controller.js:
```diff
@@ -0,0 +1,40 @@
+    if (req.preferenceMode === 'isInAppNotification') { // If request is isInAppNotification
+      await PreferenceService
+        .updatePreference({ userId: req.userData.id }, { isInAppNotification: req.value });
+      ResponseService.setSuccess(200, `Successfully updated your ${req.preferenceMode} preferences.`);
```
Regarding my previous feedback, `req.preferenceMode` is going to output one of `isInAppNotification` or `isEmailNotification` and that is not user-friendly.

A non-technical user does not know what `isInAppNotification` stands for, for example.

adesege commented on src/helpers/notifications/script.js:
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable no-console */
```
Do not disable any eslint rule

adesege commented on src/helpers/notifications/script.js:
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable no-console */
+/* eslint-disable no-undef */
+/* eslint-disable no-alert */
+const socket = io('http://localhost:3000');
```
Do not hardcode the socket host. This will not work in production.

adesege commented on src/helpers/transformErrorHandler.js:
```diff
@@ -0,0 +1,16 @@
+  if (error) {
+    const errors = [];
+    const { details } = error;
+    details.map(({ message }) => errors.push(message.split('"').join(' ')));
```
.map returns a new array. Assign a variable to the function instead of pushing to a new array.

adesege commented on src/middlewares/notification.middleware.js:
```diff
@@ -0,0 +1,19 @@
+  const { dataValues } = prefs;
+
+  const isValueSet = Object.keys(dataValues)
+    .filter((key) => key === req.preferenceMode)
```
A user can only have one preference so there is no need to filter and map. Use sequelize find method to get only one preference from the database then do your check.

victor-abz replied:
He can have them both. In DB Bernard suggested we create two columns inApp and email preference. The values are booleans.  When a user submit an update then I need to find the row to update(filter) then check its value. if it is not equal to value from body(map) that when I continue.

adesege commented on src/models/notification.js:
```diff
@@ -0,0 +1,19 @@
+export default (sequelize, DataTypes) => {
+  const Notifications = sequelize.define('Notifications', {
+    userId: DataTypes.INTEGER,
+    notificationMessage: {
```
The notification prefix is not needed.

adesege commented on src/models/trip.js:
```diff
@@ -15,5 +18,9 @@ export default (sequelize, DataTypes) => {
   };
+  Trip.afterCreate(({ dataValues }) => {
+    emitter.emit('request-created', dataValues);
+    // TripNotification.sendTripNotification(dataValues);
```
Do not leave commented code in your application. Ensure you remove all commented codes

adesege commented on src/seeders/20200106071439-demo-users.js:
```diff
@@ -0,0 +1,44 @@
+const demoUsers = {
+  up: (queryInterface, Sequelize) => queryInterface.bulkInsert('Users', [{
+    id: 31,
+    firstName: 'Manager',
```
Use Faker to generate random data

adesege commented on src/seeders/20200106072811-users-notification-mode.js:
```diff
@@ -0,0 +1,26 @@
+const demoPreferenceMode = {
```
Is there really a need to generate seeders for this task?

victor-abz replied:
Yes because each user we create in seeds needs to have default preferences set. otherwise, it would throw an error when trying to send notifications

adesege replied:
I don't see the need to generate seeders for preferences and users tables. You can manually generate these data locally. There's no point having this in staging or production environment

adesege commented on src/tests/fixtures/users.fixture.js:
```diff
@@ -78,6 +84,22 @@ export const createUser = {
   updatedAt: new Date(),
 };
+export const userWithoutManager = {
+  id: 29,
```
ids should be randomly generated

adesege commented on src/tests/notification/notification.test.js:
```diff
@@ -0,0 +1,133 @@
+        done(err);
+      });
+  });
+  it('Should not allow wrong notification mode', (done) => {
```
This test case is duplicated

adesege commented on src/tests/notification/notification.test.js:
```diff
@@ -0,0 +1,133 @@
+        done(err);
+      });
+  });
+  it('Should not accept requests with no manager', (done) => {
```
This test case is duplicated

adesege commented on src/tests/trip/multi-city-trip.test.js:
```diff
@@ -120,7 +120,7 @@ describe('Test multi city trip:', () => {
       .post('/api/multi-city-trips')
       .set('Authorization', loggedInToken)
-      .send([...multiCitytrip, { ...multiCitytrip[0], destinationId: 3 }])
+      .send([...multiCitytrip, { ...multiCitytrip[0], destinationId: 100 }])
```
`destinationId` should be randomly generated.

adesege commented on src/validations/notification.validation.js:
```diff
@@ -0,0 +1,21 @@
+import ResponseService from '../services/response.service';
+
+const validateNotification = (req, res, next) => {
+  if (Object.keys(req.body).length !== 1) {
```
Use Joi to handle the validation instead

adesege reviewed (changes requested):
Some feedback do not only apply to the line it was given. Ensure you go through your code before requesting for another review.

adesege commented on src/services/notification.service.js:
```diff
@@ -0,0 +1,53 @@
+   */
+  static async sendNotifications(receiver, notification) {
+    // Ensure notification is an object which contain notificationMessage
+    if (!users[receiver]) return;
```
Write test for this line

adesege commented on src/services/trip-notification.service.js:
```diff
@@ -0,0 +1,70 @@
+      const { isRead, updateAt, createdAt, ...notificationData } = data.dataValues;
+
+      // email nofitication
+      if (managerPreferences.isEmailNotification === true) {
```
Ensure you add a test for this line and line 63

adesege commented on src/tests/notification/socketio.test.js:
```diff
@@ -0,0 +1,110 @@
+/* eslint-disable no-unused-vars */
```
Remove this file from the test coverage report

victor-abz committed:
feature(notifications): manager can get notification

Manager can opt in to certain notification mode(email / in-app)
Manager can opt out to certain notification mode

[Finishes #169817550]

adesege commented on src/public/script.js:
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable no-console */
```
Do not disable any eslint rule

victor-abz replied:
This is the file for the front-end notification test that we said we will be deleting. Because some values are gotten in the connected Html, then it is hard for me to resolve the eslint errors

adesege replied:
Prefix each variable with `window.` because all declared variables on the browser are scoped to the window object. Moreover, that has nothing to do with the eslint rule you disabled. You shouldn't have console.log in your code by the way.

victor-abz replied:
Console.log was removed. but even if I set every variable. `window.` the error with eslint still come

adesege commented on src/public/script.js:
```diff
@@ -0,0 +1,52 @@
+  socket
+    .emit('authenticate', { token }) // send the jwt
+    .on('authenticated', () => {
+      // console.log(value);
```
Remove commented code

adesege commented on src/routes/accommodation.route.js:
```diff
@@ -1,11 +0,0 @@
-import express from 'express';
```
Why is this file deleted?

adesege commented on src/seeders/20200106071439-demo-users.js:
```diff
@@ -0,0 +1,45 @@
+
+const demoUsers = {
+  up: (queryInterface, Sequelize) => queryInterface.bulkInsert('Users', [{
+    id: 31,
```
The id should be auto-generated.

victor-abz replied:
Because I need to use the same Id in the preference seeders, 
Especially for line 5, 13, 17 If I remove the eslint rule and then try to assign variables, then the HTML won't work.

adesege replied:
Then you should look for a way to dynamically map each created user to their preference.

adesege commented on src/helpers/transformErrorHandler.js:
```diff
@@ -0,0 +1,16 @@
+  const { error } = schema.validate(body);
+
+  if (error) {
+    // const errors = [];
```
No commented code

adesege replied:
Ensure you check through the codebase and remove any commented code

adesege commented on src/middlewares/auth.middleware.js:
```diff
@@ -11,6 +11,14 @@ const authMiddleware = {
+  checkIfUserHaveManager: async (req, res, next) => {
+    const { lineManagerId } = await UserService.findUserByProperty({ email: req.userData.email });
+    if (lineManagerId === null || lineManagerId === undefined) {
+      ResponseService.setError(401, 'You don\'t have manager assigned yet. Please contact Admin');
```
The status code here shouldn't be 401.

adesege commented on src/migrations/20191218075931-create-users.js:
```diff
@@ -64,9 +64,6 @@ export function up(queryInterface, Sequelize) {
       type: Sequelize.BOOLEAN,
       defaultValue: false
     },
-    lineManagerId: {
```
Why is this column removed from migration?

victor-abz replied:
There were duplicates from develop. both in migration and model

adesege commented on src/public/script.js:
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable no-alert */
```
Really not happy I have to give this feedback the 3rd time. Do not disable any eslint rule.

adesege commented on src/services/preference.service.js:
```diff
@@ -0,0 +1,44 @@
+   * @memberof PreferenceService
+   * @return {notification} preferences
+   */
+  static findPreferences(property) {
```
There is a mismatch here. The function name is in the plural form but it's only going to return one preference for a particular user.
